File: wasm/src/TecNode.hs
{-# LANGUAGE DerivingVia #-}

module TecNode where

import Data.Aeson
  ( FromJSON,
    ToJSON (toEncoding),
    defaultOptions,
    genericToEncoding,
  )
import GHC.Generics (Generic)
import Optics.Core
data TecNodeIndex = TecNodeIndexInst String | TecNodeIndexWildcard deriving (Show, Generic)

instance ToJSON TecNodeIndex where
  toEncoding = genericToEncoding defaultOptions

instance FromJSON TecNodeIndex

data TecNodeAttribute
  = TecNodeConAttribute String
  | TecNodeTextAttribute String
  | TecNodeIntAttribute Integer
  | TecNodeFracAttribute Rational
  deriving (Show, Generic)

instance ToJSON TecNodeAttribute where
  toEncoding = genericToEncoding defaultOptions

instance FromJSON TecNodeAttribute

data TecNode = TecNode
  { indexCombination :: [TecNodeIndex],
    tecNodeAttributes :: [TecNodeAttribute]
  }
  deriving (Show, Generic)

_indexCombination :: Lens' TecNode [TecNodeIndex]
_indexCombination = lens indexCombination (\(TecNode _ ys) xs' -> TecNode xs' ys)

_tecNodeAttributes :: Lens' TecNode [TecNodeAttribute]
_tecNodeAttributes = lens tecNodeAttributes (\(TecNode xs _) ys' -> TecNode xs ys')

instance ToJSON TecNode where
  toEncoding = genericToEncoding defaultOptions

instance FromJSON TecNode

data TecNodeSet = TecNodeSet
  { tecNodeClass :: String,
    tecNodeSet :: [TecNode]
  }
  deriving (Show, Generic)

instance ToJSON TecNodeSet where
  toEncoding = genericToEncoding defaultOptions

instance FromJSON TecNodeSet

newtype TecNodeAST = TecNodeAST
  { tecNodeSets :: [TecNodeSet]
  }
  deriving (Show, Generic)

instance ToJSON TecNodeAST where
  toEncoding = genericToEncoding defaultOptions

instance FromJSON TecNodeAST

File: web/libs/codec/src/TecNode.ts
import { Schema as S } from "effect";
import * as E from "effect";
import * as TecClass from "./TecClass.js";
import * as TecEnum from "./TecEnum.js";

export const TecNodeIntAttribute = S.Struct({
  tag: S.tag("TecNodeIntAttribute"),
  contents: S.Number.pipe(S.int()),
});

export const TecNodeFracAttribute = S.Struct({
  tag: S.tag("TecNodeFracAttribute"),
  contents: S.Struct({
    numerator: S.Number.pipe(S.int()),
    denominator: S.Number.pipe(S.int()),
  }),
});

export const TecNodeTextAttribute = S.Struct({
  tag: S.tag("TecNodeTextAttribute"),
  contents: S.String,
});

export const TecNodeConAttribute = S.Struct({
  tag: S.tag("TecNodeConAttribute"),
  contents: S.String,
});

const TecNodeAttribute = S.Union(
  TecNodeIntAttribute,
  TecNodeFracAttribute,
  TecNodeTextAttribute,
  TecNodeConAttribute,
);

export type TecNodeAttribute = typeof TecNodeAttribute.Type;
export const TecNodeIndexWildcard = S.Struct({
  tag: S.tag("TecNodeIndexWildcard"),
});
export const TecNodeIndexInst = S.Struct({
  tag: S.tag("TecNodeIndexInst"),
  contents: TecEnum.TecEnumValue,
});
export type TecNodeIndexWildcard = typeof TecNodeIndexWildcard.Type;
export type TecNodeIndexInst = typeof TecNodeIndexInst.Type;
const TecNodeIndex = S.Union(TecNodeIndexWildcard, TecNodeIndexInst);
export const tecNodeIndexEquivalence = S.equivalence(TecNodeIndex);
export type TecNodeIndex = typeof TecNodeIndex.Type;
export type TecNodeIndexSet = E.Array.NonEmptyReadonlyArray<TecNodeIndex>;
const IndexCombination = S.NonEmptyArray(TecNodeIndex);
export type IndexCombination = typeof IndexCombination.Type;
export const TecNode = S.Struct({
  indexCombination: IndexCombination,
  tecNodeAttributes: S.Array(TecNodeAttribute),
});
export type TecNode = typeof TecNode.Type;

export const TecNodeSet = S.Struct({
  tecNodeClass: TecClass.TecClassName,
  tecNodeSet: S.NonEmptyArray(TecNode),
});
export type TecNodeSet = typeof TecNodeSet.Type;
type RNE<T> = E.Array.NonEmptyReadonlyArray<T>;
export const lens = {
  tecNodeSet: {
    over: {
      tecNodeSet(f: (tecNodeSet: RNE<TecNode>) => RNE<TecNode>) {
        return function (tecNodeSet: TecNodeSet): TecNodeSet {
          return TecNodeSet.make({
            tecNodeClass: tecNodeSet.tecNodeClass,
            tecNodeSet: f(tecNodeSet.tecNodeSet),
          });
        };
      },
    },
  },
};

export const TecNodeAST = S.Struct({ tecNodeSets: S.Array(TecNodeSet) });
export type TecNodeAST = typeof TecNodeAST.Type;
